[UnifiedPDF] Cursor updates are incorrect over text/image elements for untagged PDFs.

aprotyas opened the pull request:
#### bcb2f892a5d1c79d21b6df3222186a81f014e059
<pre>
[UnifiedPDF] Cursor updates are incorrect over text/image elements for untagged PDFs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265908">https://bugs.webkit.org/show_bug.cgi?id=265908</a>
<a href="https://rdar.apple.com/119217538">rdar://119217538</a>

Reviewed by Tim Horton.

The CGPDFPageLayoutGetAreaOfInterestAtPoint interface does not correctly
identify text/image elements on Untagged PDF files. This made our hit
test results incomplete on said PDF files. To address this issue, we
adopt new PDFPage API `areaOfInterestAtPoint`, which not only hit tests
for annotations, but also for text and image on a page, thus subsuming
both -[PDFPage annotationAtPoint:] and CGPDFPageLayoutGetAreaOfInterestAtPoint
while maintaining correctness across tagged and untagged PDFs.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

Canonical link: <a href="https://commits.webkit.org/275287@main">https://commits.webkit.org/275287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5a5e065cd9d63be601cfaa59b95e6c2426a7555

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41425 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43991 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37517 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17769 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34243 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41999 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17369 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35671 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14904 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15085 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45343 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37615 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36994 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40742 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16240 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13321 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39144 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17859 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9284 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17913 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17503 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1624,6 +1624,38 @@ static BOOL annotationIsLinkWithDestination(PDFAnnotation *annotation)
+            pdfElementTypes.add(PDFElementType::Annotation);
+
+        if (areaOfInterest & kPDFLinkArea)
+            pdfElementTypes.add(PDFElementType::Link);
```
It seems like this loses the "with destination" part of annotationIsLinkWithDestination

aprotyas replied:
Good catch, adding it back.

aprotyas replied:
Fixed in [d5a5e06](https://github.com/WebKit/WebKit/commit/d5a5e065cd9d63be601cfaa59b95e6c2426a7555)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1624,6 +1624,38 @@ static BOOL annotationIsLinkWithDestination(PDFAnnotation *annotation)
+        if (areaOfInterest & kPDFControlArea)
+            pdfElementTypes.add(PDFElementType::Control);
+
+        if (areaOfInterest & kPDFTextFieldArea)
```
What about the readonly check?

aprotyas replied:
PDFKit handles it!

aprotyas committed:
[UnifiedPDF] Cursor updates are incorrect over text/image elements for untagged PDFs.
https://bugs.webkit.org/show_bug.cgi?id=265908
rdar://119217538

Reviewed by Tim Horton.

The CGPDFPageLayoutGetAreaOfInterestAtPoint interface does not correctly
identify text/image elements on Untagged PDF files. This made our hit
test results incomplete on said PDF files. To address this issue, we
adopt new PDFPage API `areaOfInterestAtPoint`, which not only hit tests
for annotations, but also for text and image on a page, thus subsuming
both -[PDFPage annotationAtPoint:] and CGPDFPageLayoutGetAreaOfInterestAtPoint
while maintaining correctness across tagged and untagged PDFs.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

Canonical link: https://commits.webkit.org/275287@main
